File: quicksort.py
def quicksort(arr: list[int]) -> list[int]:
    """Sort a list of integers using the quicksort algorithm.

    Selects the middle element as the pivot, then recursively sorts
    the sub-lists of elements less than and greater than the pivot.

    Args:
        arr: The list of integers to sort.

    Returns:
        A new sorted list in ascending order.
    """
    if len(arr) <= 1:
        return arr

    pivot = arr[len(arr) // 2]
    left = [x for x in arr if x < pivot]
    middle = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]

    return quicksort(left) + middle + quicksort(right)


if __name__ == "__main__":
    input_arr = [3, 6, 8, 10, 1, 2, 1]
    print("Input: ", input_arr)
    print("Sorted:", quicksort(input_arr))

File: quicksort.ts
/**
 * Sorts an array of numbers using the quicksort algorithm.
 *
 * Selects the middle element as the pivot, then recursively sorts
 * the sub-arrays of elements less than and greater than the pivot.
 *
 * @param arr - The array of numbers to sort.
 * @returns A new sorted array in ascending order.
 */
function quicksort(arr: number[]): number[] {
  if (arr.length <= 1) {
    return arr;
  }

  const pivot = arr[Math.floor(arr.length / 2)];
  const left = arr.filter((x) => x < pivot);
  const middle = arr.filter((x) => x === pivot);
  const right = arr.filter((x) => x > pivot);

  return [...quicksort(left), ...middle, ...quicksort(right)];
}

const input = [3, 6, 8, 10, 1, 2, 1];
console.log("Input: ", input);
console.log("Sorted:", quicksort(input));
